### DISCUSSION TITLE
Using const() in different classes

### DISCUSSION THREAD

**Discussion by fmafma** [micropython/micropython#16075] [General]
I found a strange behaviour, probably due to my lack of const() understanding.

If the test.py module contains:

```
class A:
    C = const(1)

class B:
    C = const(2)
```
and I try to import this module, I get:
```
>>> import test
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/lib/test.py", line 4, in B
SyntaxError: can't assign to expression
```
What I did wrong?

**Comment by fmafma** [comment]
Also, if I type this:
```
>>> CONST_X = const(123)
>>> CONST_Y = const(2 * CONST_X + 1)
```
I get:
```
Traceback (most recent call last):
  File "<stdin>", line 1
SyntaxError: not a constant
```
But if I put these 2 lines in a module and import that module, it works!

**Comment by robert-hh** [comment]
When you assign a value to a symbol using `const()` the compiler treats that symbol as a constant value, just like a number. These symbols are global in a script, not local to a class to function. So the second assignment it treated by the compiler like `1 = const(2)`.

**Reply by robert-hh** [reply]
In REPL compiler constants are not memorized between command lines. Each line is a separate call of the compiler. So in the second line the value of CONST_X is lost.

**Comment by fmafma** [comment]
Ok, I understand for the first example. So const() is more like creating an alias/define? (the documentation is not very clear about that)

But I don't understand your second answer:
```
>>> A = const(1)
>>> 1
```
So it still exists between the 2 calls. So why it can't use it in the next const() definition?

**Reply by robert-hh** [reply]
Seems you#re right. Then it may be that const symbols must not be used in an expressing when using REPL.

**Comment by fmafma** [comment]
Mmm, yes and no:
```
>>> a = const(1)
>>> b = a + 1
>>> c = sqrt(a + 1)
>>> d = const(a + 1)
Traceback (most recent call last):
  File "<stdin>", line 1
SyntaxError: not a constant
```
It only fails when used in const()...

**Comment by fmafma** [comment]
Another strange thing:
```
>>> const(2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'const' isn't defined
>>> a = const(2)
>>>
```
I think I will stay away from const() for now ;o)

**Comment by peterhinch** [comment]
`const` is a directive to the Python compiler and doesn't really make sense at the REPL. If in a module you have
```py
_WIDTH = const(256)
```
and later in your code you have
```py
def foo(x):
    y  = x + _WIDTH
```
the bytecode compiler use the number 256 instead of looking up a global variable. Saves time and code.

**Comment by fmafma** [comment]
Ok, but why does it have to become global? Why not follow the Python object hierarchy?
As I said, I see it more as a define... I find confusing the way it is implemented.
